daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,163 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dsom.walker
+
+import org.apache.daffodil.dpath.NodeInfo.PrimType
+
+trait CommonContextView {
+  def namespaces: scala.xml.NamespaceBinding
+}
+
+trait TermView extends CommonContextView {
+  def isArray: Boolean
+  def isOptional: Boolean
+  def walkDSOM[T](walker: AbstractDSOMWalker[T]): Unit = {
+    walker.onTermBegin(this)
+    walker.onTermEnd(this)
+  }
+}
+
+trait SimpleTypeView {
+  def primType: PrimType
+}
+trait ComplexTypeView {
+  def group: ModelGroupView
+}
+
+trait ModelGroupView extends TermView {
+  def groupMembers: Seq[TermView]
+  override def walkDSOM[T](walker: AbstractDSOMWalker[T]): Unit = {
+    walker.onTermBegin(this)
+    this.groupMembers.foreach(_.walkDSOM(walker))
+    walker.onTermEnd(this)
+  }
+}
+
+trait ChoiceView extends ModelGroupView
+trait SequenceView extends ModelGroupView
+trait GroupRefView extends ModelGroupView {
+  def isHidden: Boolean
+  override def walkDSOM[T](walker: AbstractDSOMWalker[T]): Unit = if (!isHidden) super.walkDSOM(walker)
+}
+
+trait ElementDeclView extends CommonContextView {
+  def isSimpleType: Boolean
+  def isComplexType: Boolean
+  def simpleType: SimpleTypeView
+  def complexType: ComplexTypeView
+  def hasDefaultValue: Boolean
+}
+
+trait ElementBaseView extends ElementDeclView with TermView {
+  def name: String
+  override def walkDSOM[T](walker: AbstractDSOMWalker[T]): Unit = {
+    walker.onTermBegin(this)
+    if (isComplexType) {
+      walker.onTypeBegin(Right(complexType))
+      complexType.group.walkDSOM(walker)
+      walker.onTypeBegin(Right(complexType))
+    } else {
+      walker.onTypeBegin(Left(simpleType))
+      walker.onTypeBegin(Left(simpleType))
+    }
+    walker.onTermEnd(this)
+  }
+
+}
+
+trait RootView extends ElementBaseView
+
+/**
+ * A class designed to walk the internal representation of a DFDL Schema File.
+ *
+ * There are 2 main event handlers an implementing class has to worry about: one for Terms, and
+ * another for Types.  These are called as the DSOM is walked.
+ *
+ * Though recursion is used here to define the walk, it is not advised to use recursion between
+ * these event handlers.  Instead, consider a stack-like structure, as the DFDL Schema structure
+ * as well as the recursive method call structure can be represented by trees.
+ * @tparam T the return type of walkDSOMSchema and onWalkEnd.  This should be a class used to
+ *           represent the structure of the Schema.  You could always make it Unit if you don't
+ *           want these methods to return anything.
+ */
+abstract class AbstractDSOMWalker[T] {
+
+  /**
```

Comment:
Regarding the *View stuff, this seems reasonable to me. One concern is that we probably want this to be a public stable API, that works in both Java and Scala. I *think* Scala traits get turned into interfaces that a Java class can access, but I'm not 100% sure. If we generate JavaDocs, which we probably want to do, I'm also not sure if the Scala traits turn into something useful that makes sense for a java API. Note that we use the genjavadocs plugin to generate javadocs for daffodil-japi. I'm wondering if we want to do something similar? So maybe we want to move all the walker related things that are part of the public API (i.e. everything in the dsom.walker package) to a daffodil-dsom-walker subproject? Then we can generate JavaDocs for that just like with do with the Daffodil JAPI.

Related, I'm looking into [DAFFODIL-2368](https://issues.apache.org/jira/browse/DAFFODIL-2368), which is related to how we can merge docs from multiple subprojects. I think looking at the Javadocs that results from this would be a clear way to view what parts of this API are public and what aren't.

Reply:
While we do want stability, I think we should separate that from having a need for Java support. This particular API is so close to the core of the Daffodil schema compiler that I think requiring it to be scala only is fine. I don't think it is necessary for people to be able to write extensions that walk the internals parse trees in Java. 

It's not that this is not possible, it's just something I think adds little value for us to support, at least for now with this being the very first Daffodil "metadata integration".

Reply:
One concern with requireing scala is building a nar for a NiFi controller/processer uses a maven archetype template to build the project structure. If the API requires Scala, building such a nar might be problematic, or at least this generated maven project needs to be modified support scala compilation. I assume that's possible, but I'm not sure. Other integrations might have similar issues.

And if the the utltimate goal is to upstream this into NiFi tree, it almost certainly needs to be Java friendly--as far as I know, NiFi doesn't have any Scala code in their source tree.

It we decide to only support scala (which doesn't seem unreasonable), and if there's a potential for changes to support Java, I would at least recomend we say that this API is experiemental and subject to change so that we don't have to worry as much about backwards compatibility and can more easily make such changes.

Reply:
Expermental until we have a second workable metadata integration sounds like the right thing.

Reply:
It is possible to build nars using scala code, I previously had all of this code in a maven project that mixed NIFI-specific Java code and Scala code.  It only required a few additional packages to include.

That being said, if the NiFi developer doesn't have to code in Scala, they don't have to worry about any of that - they would only need to do this configuration if they wanted to implement AbstractDSOMWalker in Scala